samza-core/src/main/java/org/apache/samza/execution/StreamEdge.java

```diff
@@ -60,12 +69,17 @@ void addTargetNode(JobNode targetNode) {
     targetNodes.add(targetNode);
   }
 
-  public StreamSpec getStreamSpec() {
-    if (partitions == PARTITIONS_UNKNOWN) {
-      return streamSpec;
-    } else {
-      return streamSpec.copyWithPartitionCount(partitions);
+  StreamSpec getStreamSpec() {
+    StreamSpec spec = (partitions == PARTITIONS_UNKNOWN) ?
+        streamSpec : streamSpec.copyWithPartitionCount(partitions);
+
+    if (isIntermediate) {
+      ApplicationConfig appConfig = new ApplicationConfig(config);
+      if (appConfig.getAppMode() == ApplicationConfig.ApplicationMode.BATCH && appConfig.getRunId() != null) {
+        spec = spec.copyWithPhysicalName(spec.getPhysicalName() + "-" + appConfig.getRunId());
```

Comment:
This logic is somewhat similar to the logic added to JavaStreamConfig. 

Is there any way to put it in a central place? 

Not sure if I like this idea, but perhaps a BatchStreamSpec class that either extends StreamSpec (like KafkaStreamSpec does) or just provides some static methods. Thoughts?